ABNF cleanup

robUx4 opened the pull request:
This PR sits on top of #324 and #329

Follow each commit for more details on the progression of changes that lead to the removal of some of the elements in the ABNF.

The ABNF has also been reordered so that each item is defined after all the elements using it are defined.

Fixes #326

robUx4 commented:
After this Matroska will have pathes recursive pathes like this:

```
\Segment\Chapters\EditionEntry(\ChapterAtom)
\Segment\Chapters\EditionEntry(\ChapterAtom)\ChapterUID

\Segment\Tags\Tag(\SimpleTag)
\Segment\Tags\Tag(\SimpleTag)\TagName
```

There is no min or max level involved.

robUx4 commented on specification.markdown:
```diff
@@ -386,14 +386,13 @@ The path defines the allowed storage locations of the EBML Element within an EBM
-EBMLParentPath        = EBMLFixedParent EBMLLastParent
-EBMLFixedParent       = *(EBMLPathAtom)
+EBMLFullPath          = [EBMLParentPath] [EBMLGlobalParent] 
+                        EBMLElementPath
```
You understand correctly.

dericed replied:
OK, so IIUC an EBMLFullPath can be an EBMLParentPath followed by an EBMLElementPath with no EBMLGlobalParent, but in that case how could the boundary between the EBMLParentPath and the EBMLElementPath be determined?

Or is EBMLParentPath only relevant to elements with an EBMLGlobalParent? If so, then  EBMLParentPath could be added to EBMLGlobalParent rather than a component of EBMLFullPath.

robUx4 replied:
EBMLParentPath is empty or multiple `\abcdefg` parts
EBMLElementPath is non-empty and a `\abcdefg` part.
The EBMLElementPath is the last `\abcdefg` of a full path. The rest is the EBMLParentPath. That's how it is deterministic.

The EBMLGlobalParent is always in the form `(x*y)` which is distinct enough from a EBMLPathAtomRecursive which is in the form `(\abcdef)`.

Void/CRC-32 still include the `\` in the EBMLGlobalParent part. I think it makes it easier to read that it's the repetition of some path, so I'll add it back in EBMLGlobalParent.

robUx4 commented on specification.markdown:
```diff
@@ -386,14 +386,13 @@ The path defines the allowed storage locations of the EBML Element within an EBM
-EBMLPathAtomRecursive = "(1*(" EBMLPathAtom "))"
-EBMLLastParent        = EBMLPathAtom / EBMLGlobalParent
-EBMLGlobalParent      = "(" GlobalParentOccurence "\)"
+EBMLGlobalParent      = "(" GlobalParentOccurence ")"
```
I would prefer not to. Because changing this sentence would be tricky:

> The GlobalParentOccurence part is interpreted as an ABNF Variable Repetition

While `x*y` is a valid repetition in ABNF, `"("x*y")"` or `(x*y)` are not.
Many of the items are split so we can define them in the text.

robUx4 commented:
These new commits add more cleaning and a differentiation of the last PathAtom so it can be referenced explicitly in the text.

robUx4 commented:
And also used `GlobalPlaceholder` as it's not a real path like the other parts. Just a virtual placeholder for any PathAtom

robUx4 commented:
I added an extra commit in https://github.com/cellar-wg/ebml-specification/tree/global-in-children to add the GlobalPlaceholder in children of a Global Element.

dericed commented on specification.markdown:
```diff
@@ -386,14 +386,14 @@ The path defines the allowed storage locations of the EBML Element within an EBM
-EBMLGlobalParent      = "(" GlobalParentOccurence "\)"
+GlobalPlaceholder     = "(" GlobalParentOccurence "\)"
 EBMLAtomName          = 1*(EBMLNameChar)
 EBMLNameChar          = ALPHA / DIGIT / "-" / "."
```
EBMLNameChar doesn't match the spec. I think it should be something like
`EBMLNameChar = ALPHA / DIGIT 0*[ALPHA / DIGIT / "-" / "."]`

robUx4 replied:
Actually the `EBMLAtomName` start with be a `ALPHA / DIGIT`. `EBMLNameChar` is the list of possible chars after the first one. But good catch.

dericed commented on specification.markdown:
```diff
@@ -404,23 +404,25 @@ PathMaxOccurrence     = 1*DIGIT ; no upper limit
 The starting PathDelimiter of the path corresponds to the root of the EBML Document.
 
-In some cases the EBMLLastParent part of the path is an EBMLGlobalParent. A path with a EBMLGlobalParent defines a Global Element; see (#global-elements). Any path that starts with the EBMLFixedParent of the Global Element and matches the occurrences found in the GlobalParentOccurence is a valid path for the Global Element. 
+If the EBMLLastPath is an EBMLPathAtomRecursive, the EBML Element can occur within itself recursively (see (#recursive)).
```
I'd suggest that this sentence should immediately follow the one on line 407, such as

>The EBMLAtomName part of the EBMLLastPath MUST be equal to the `@name` attribute of the EBML Schema. If the EBMLLastPath is an EBMLPathAtomRecursive, the EBML Element can occur within itself recursively (see (#recursive)).

robUx4 replied:
OK

robUx4 commented:
This Matroska branch reflects the changes in this branch.
https://github.com/cellar-wg/matroska-specification/tree/del-path-occurences

Once this branch is merged, that branch should be merged too.

dericed commented on specification.markdown:
```diff
@@ -404,23 +404,24 @@ PathMaxOccurrence     = 1*DIGIT ; no upper limit
+The repetition amounts correspond to the amount of unspecified valid Elements that can be found in place of the GlobalPlaceholder for the path to be valid.
 
-As an example, a `path` of `1*(\Segment\Info)` means the element Info is found inside the Segment elements at least once and with no maximum iteration. An element SeekHead with path `0*2(\Segment\SeekHead)` may not be found at all in its Segment parent, once or twice but no more than that.
+PathMinOccurrence represents the minimum number of element path required between the EBMLParentPath and the EBMLLastPath. For example 0 means the EBMLLastPath can be right after the EBMLParentPath, 1 means there has to be at least an element between the EBMLParentPath and the EBMLLastPath. If PathMinOccurrence is not present then that GlobalPlaceholder has a PathMinOccurrence value of 0.
```
`the minimum number of element path` could be better worded. It feels unclear to me. 

Actually the first sentence is hard to process. For instance with `/Root/(2*)/global` if PathMinOccurrence=2 and there needs to be at least 2 `number of element path required between the EBMLParentPath and the EBMLLastPath`. Does that mean?

```
/Root
/Root/ChildA
/Root/ChildB
/Root/global
```

or
```
/Root/ChildA/ChildB/global
```

Under different interpretations ChildA and ChildB are `between` Root and global.

robUx4 replied:
That's `\Root\ChildA\ChildB\global`. I'll see if I can find something bettter.

robUx4 replied:
Added some more detailed example, hopefully not too verbose.

robUx4 commented:
Side note, we define `PathMaxOccurrence` but neither in EBML or Matroska it has a different value than the default one. We could get rid of it. Not sure there are cases where semi-global elements make sense.

dericed commented on specification.markdown:
```diff
@@ -404,23 +404,36 @@ PathMaxOccurrence     = 1*DIGIT ; no upper limit
+The repetition amounts correspond to the amount of unspecified valid Elements that can be found in place of the GlobalPlaceholder for the path to be valid.
 
-PathMaxOccurrence represents the maximum number of element path possible between the EBMLFixedParent and the Global Element EBMLElementPath. It cannot have the value 0 as it would be the Global Element can only be found right after the EBMLFixedParent, in which case it's not a Global Element anymore. If PathMaxOccurrence is not present then there is no upper bound for the permitted number of occu ...
+PathMinOccurrence represents the minimum number of path elements required between the EBMLParentPath and the EBMLLastPath.
```
But what does this mean if there is no EBMLParentPath (it is optional)?

robUx4 replied:
Yes it's optional. And in the only cases it's used CRC-32 and Void it is empty. In that case it's the root.
Maybe we can mention that an empty `EBMLParentPath` means it's the root.

Alternatively we can rework the ABNF so that the root is always present. In that case the EBMLLastPath would not contain the `\` anymore.

dericed replied:
Hi @robUx4, I tried with the latter option and making EBMLParentPath mandatory, where that can be either `\` as Root or the series of intermediate paths. Please review.

dericed commented on specification.markdown:
```diff
@@ -404,23 +404,36 @@ PathMaxOccurrence     = 1*DIGIT ; no upper limit
+PathMaxOccurrence cannot have the value 0 as it would mean the EBMLLastPath can only be found right after the EBMLParentPath, in which case it's not a Global Element anymore and a GlobalPlaceholder MUST NOT be used. 
+
+For example consider an EBML Path `\a(0*1\)\global`, there has to be at most one path element between the EBMLParentPath `\a` and the EBMLLastPath `\global`.
+So the `global` EBML Element can be found inside the `\a` EBML Element or inside the `\a\b` EBML Element.
```
Like the comment above, I'm not sure how to understand this if it was `(0*1\)\global`.

robUx4 commented on specification.markdown:
```diff
@@ -968,7 +968,7 @@ A Global Element can be found in many Parent Elements, allowing the same number
 name: CRC-32
 
-path: `(1*\)\CRC-32`
+path: `\(1*\)\CRC-32`
```
The idea is that the parent `\` is not part of the element name anymore, ie `\(1*\)CRC-32` or `\(*\)Void`. So if you read it like a pattern for CRC-32 you need at least one extra \ in the path. For Void it can have 0 (or plenty) extra \ in the path.

robUx4 commented:
I reworked the notation for recursive elements (just a +, similar to what was originally used in the Matroska specs).

I also removed the \ from the global element name after the `(*\)` so the number of \ corresponds to the amount of levels you can find the element in. (not referred to as levels)

I redid the occurence test as the ABNF analogy was causing more problem than anything. It's not a repetition as found in ABNF but a range of possible occurences, of elements replacing the global placeholder.

I added a text file with some path examples to show the various possibilities and then ABNF parts they are decomposed into. That really helps finding valid/invalid syntaxes (or reducing the recursive to just a plus).

dericed commented on specification.markdown:
```diff
@@ -386,41 +386,60 @@ The path defines the allowed storage locations of the EBML Element within an EBM
+EBMLAtomName          = ALPHA / DIGIT 0*EBMLNameChar
+EBMLNameChar          = ALPHA / DIGIT / "-" / "."
+
+GlobalPlaceholder     = "(" GlobalParentOccurence "\)"
```
Since parenthesis are no longer used for recursion, I suggest a cleaner look and use:

`GlobalPlaceholder     = "(" GlobalParentOccurence ")"`

dericed replied:
I have a slight preference for:

`GlobalPlaceholder = "(" GlobalParentOccurence ")" PathDelimiter`

As in, instead of `\a\(0-1\)global` to have `\a\(0-1)\global`

robUx4 replied:
But that's the point. (*\) means there are plenty of abc\def\ghi\ in place of the placeholder. The \ is "repeated" for each occurence. If you leave it out IMO it's less readable.

dericed commented on specification.markdown:
```diff
@@ -386,41 +386,60 @@ The path defines the allowed storage locations of the EBML Element within an EBM
 
-The GlobalParentOccurence part is interpreted as an ABNF Variable Repetition. The repetition amounts correspond to the amount of unspecified Parent Element levels there can be between the EBMLFixedParent and the actual EBMLElementPath.
+A path with a GlobalPlaceholder as the EBMLLastParent defines a Global Element; see (#global-elements).
+If the element has no EBMLLastParent part or the EBMLLastParent part is not a GlobalPlaceholder then the Element is not a Global Element.
```
Maybe more logical to say `If the element has no EBMLParents part or...`, but still confused.

Suppose you have a Global Master element like:

`/(0-1)/GlobalMaster`

then a Child Element of that would be:
`/(0-1)/GlobalMaster/SubElement`

With this definition, `SubElement` is not considered as a Global Element. Is that intended?

robUx4 replied:
That's correct. The SubElement doesn't get to be placed within all kinds of various unknown parents. It has only a fixed parent.

dericed commented on specification.markdown:
```diff
@@ -386,41 +386,60 @@ The path defines the allowed storage locations of the EBML Element within an EBM
+The `@path` value MUST be unique within the EBML Schema. The `@id` value corresponding to this `@path` MUST NOT be defined for use within another EBML Element with the same EBMLParentPath as this `@path`.
 
-The GlobalParentOccurence part is interpreted as an ABNF Variable Repetition. The repetition amounts correspond to the amount of unspecified Parent Element levels there can be between the EBMLFixedParent and the actual EBMLElementPath.
+A path with a GlobalPlaceholder as the EBMLLastParent defines a Global Element; see (#global-elements).
```
But what about a path with a GlobalPlaceholder as an IntermediatePathAtom?

robUx4 replied:
(One of) its parent is a global element, but that element in particular is not a global element. The global "placement" only applies to that parent.

dericed commented on specification.markdown:
```diff
@@ -386,41 +386,60 @@ The path defines the allowed storage locations of the EBML Element within an EBM
 
-PathMaxOccurrence represents the maximum number of element path possible between the EBMLFixedParent and the Global Element EBMLElementPath. It cannot have the value 0 as it would be the Global Element can only be found right after the EBMLFixedParent, in which case it's not a Global Element anymore. If PathMaxOccurrence is not present then there is no upper bound for the permitted number of occu ...
+If PathMinOccurrence is not present then that GlobalParentOccurence has a PathMinOccurrence value of 0.
+If PathMaxOccurrence is not present then there is no upper bound for the permitted amount of EBMLPathAtom possible to replace the GlobalPlaceholder.
```
I think we could use a statement, like `PathMaxOccurrence MUST be greater than PathMinOccurrence.`

robUx4 replied:
Yeah I left it out because it seems obvious but why not.
It can't be done in the ABNF itself ;(

dericed commented on specification.markdown:
```diff
@@ -947,15 +966,15 @@ description: The version of the DocTypeExtension. Different DocTypeExtensionVers
 name: CRC-32
 
-path: `*1((1*\)\CRC-32)`
+path: `\(*\)CRC-32`
```
Needs an update, here an in Void, as we swapped from `*` to `-`.

robUx4 replied:
I removed the `-` change. It slipped in the push but I didn't want it in the end. `-` is part of Element names so it might be confusing an ABNF parser.

robUx4 replied:
Never mind. I cannot be confused with a name as it's in parenthesis. It's easier to display a number range with `-` so I put it back.

dericed commented:
47 commits !_!
should we squash-merge?

dericed commented:
and if i understand correctly, abnf-examples.txt is simply an example doc for the repository and not part of the RFC, right?

robUx4 commented:
Yes just an example for us. We may put it as an annex if needed.

I prefer to keep the history of commits (and the reasons for the changes) so just rebase and merge.
